### PR TITLE
IoT Core, Provisioning state APIs implementations

### PR DESCRIPTION
**Pull request by CIPop**
- Implementations for az_iot_core state and retry public APIs + UT.
- Adding operation_state parsing APIs + UT + updated sample.
- Clarifying naming for Provisioning APIs (breaking API change).
- Removing some az_iot_core functions from the public API surface.
- Fixing Provisioning invalid JSON parsing bug + UT.

Doc:
- Updating PUML (switching to svg rendering).
- Changing from @note to @remark for non-important items (leaving @note for paragraphs that _must_ be understood, @remarks for clarifications).

### CODE REVIEW

**Comment by CIPop** [comment]
@gilbertw @JeffreyRichter @ericwol-msft PTAL at public API changes (.h files)

**Comment by CIPop** [comment]
Master got updated and broke this.

File: sdk/iot/provisioning/inc/az_iot_provisioning_client.h [hunk 1]
```diff
@@ -228,16 +233,19 @@ typedef struct az_iot_provisioning_client_registration_state
 typedef struct az_iot_provisioning_client_register_response
 {
   az_iot_status status; /**< The current request status.
-                         * @note The authoritative response for the device registration operation
+                         * @remark The authoritative response for the device registration operation
                          * (which may require several requests) is available only through
-                         * #registration_state.  */
+                         * #operation_status.  */
```
Comment (danewalton) [inline sdk/iot/provisioning/inc/az_iot_provisioning_client.h:238]: I think this should be backticks. `operation_status` isn't a type and won't resolve.

File: sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c [hunk 2]
```diff
@@ -300,42 +312,42 @@ static int get_operation_status()
...
       }
-
-      is_operation_complete = true;
     }
 
     MQTTClient_freeMessage(&message);
     MQTTClient_free(topic);
   }
-
+
```
Comment (danewalton) [inline sdk/iot/provisioning/samples/src/paho_iot_provisioning_example.c:350]: unnecessary space?

File: sdk/iot/provisioning/src/az_iot_provisioning_client.c [hunk 3]
```diff
@@ -335,6 +335,11 @@ AZ_INLINE az_result _az_iot_provisioning_client_payload_registration_state_parse
     }
   }
 
+  if (found_assigned_hub != found_device_id)
```
Comment (danewalton) [inline sdk/iot/provisioning/src/az_iot_provisioning_client.c:338]:
```suggestion
  if (!(found_assigned_hub && found_device_id))
```
having not equals makes me believe `false == false` would pass here. I have to look back up at the implementation to see if that is caught or not.

Comment (danewalton) [reply sdk/iot/provisioning/src/az_iot_provisioning_client.c:338]: or even have `if (found_assigned && found_device_id)` and then after default to `AZ_ERROR_ITEM_NOT_FOUND`

File: sdk/iot/core/inc/az_iot_core.h [hunk 4]
```diff
@@ -58,56 +62,49 @@ typedef enum
...
+ * @param[in] min_retry_delay_msec The minimum time, in milliseconds, to wait before a retry.
  * @param[in] max_retry_delay_msec The maximum time, in milliseconds, to wait before a retry.
  * @param[in] random_msec A random value between 0 and the maximum allowed jitter, in milliseconds.
  * @return The recommended delay in milliseconds.
  */
 AZ_NODISCARD int32_t az_iot_retry_calc_delay(
     int32_t operation_msec,
     int16_t attempt,
-    int32_t retry_delay_msec,
+    int32_t min_retry_delay_msec,
```
Comment (gilbertw) [inline sdk/iot/core/inc/az_iot_core.h:102]: Why split min and max?  Guarantees to never be late are extremely hard to enforce due to schedulers.

File: sdk/iot/core/inc/az_iot_core.h [hunk 5]
```diff
@@ -120,7 +117,7 @@ AZ_NODISCARD int32_t az_iot_retry_calc_delay(
  * occurrence of (but not including the) `delimiter`, or the end of `source` if `delimiter` is not
  * found. If `source` is empty, AZ_SPAN_NULL is returned instead.
  */
-AZ_NODISCARD az_span az_span_token(az_span source, az_span delimiter, az_span* out_remainder);
+AZ_NODISCARD az_span _az_span_token(az_span source, az_span delimiter, az_span* out_remainder);
```
Comment (gilbertw) [inline sdk/iot/core/inc/az_iot_core.h:120]: Based on the method name this should be in az_span.  This also sounds like it might be a json parser api.  Is there a reason for not adding to az_span or using the json parser?

Comment (CIPop) [reply sdk/iot/core/inc/az_iot_core.h:120]: That's right and the reason I moved it to internal - it should definitely not be a public API within the IoT headers.
There was some discussion a while back (@JeffreyRichter @ewertons have more context) and it was decided to remain outside of the span API - not sure why.

I do think the token API is useful: 1. we used it and liked it, 2. C's StdLib provides `strtok`

Comment (CIPop) [reply sdk/iot/core/inc/az_iot_core.h:120]: I will defer this to offline discussions and a later PR.

File: sdk/iot/core/inc/az_iot_core.h (hunk 4 above)
Comment (CIPop) [reply sdk/iot/core/inc/az_iot_core.h:102]: The API is using the internal function provided by AZ SDK which has retry_delay_msec defined as the minimum retry time.
If found that confusing so I renamed the arg in our public API.

I don't understand the timing enforcement guarantee problem. This API is implementing an exponential back-off with jitter. It's very common with distributed (network) protocols to implement a randomized back-off. The exponential aspect is in-place to avoid servers being DDoSed by devices.

More details on how this should be used here: https://github.com/Azure/azure-sdk-for-c/blob/master/sdk/iot/doc/mqtt_state_machine.md#retry-timing

File: sdk/iot/core/inc/az_iot_core.h [hunk 6]
```diff
@@ -21,12 +21,16 @@
...
+ * @note You MUST NOT use any symbols (macros, functions, structures, enums, etc.)
+ * prefixed with an underscore ('_') directly in your application code. These symbols
+ * are part of Azure SDK's internal implementation; we do not document these symbols
+ * and they are subject to change in future versions of the SDK which would break your code.
  */
+
 enum
 {
-  AZ_CLIENT_DEFAULT_MQTT_CONNECT_PORT = 8883,
-  AZ_CLIENT_DEFAULT_MQTT_CONNECT_CLEAN_SESSION = 0x20,
```
Comment (ericwol-msft) [inline sdk/iot/core/inc/az_iot_core.h:28]: why was this removed, I was using it?

File: sdk/iot/provisioning/src/az_iot_provisioning_client.c (hunk 3 above)
Comment (CIPop) [reply sdk/iot/provisioning/src/az_iot_provisioning_client.c:338]: > having not equals makes me believe false == false would pass here

And you would be correct :). C doesn't have XOR. false, false and true,true are the only acceptable values: IoT Hub either sends the hub and device or _none at all_ and instead sends the error code.

File: sdk/iot/provisioning/inc/az_iot_provisioning_client.h (hunk 1 above)
Comment (CIPop) [reply sdk/iot/provisioning/inc/az_iot_provisioning_client.h:238]: Right: I'm going to fix all of them (searching for # in generated html).

File: sdk/iot/core/inc/az_iot_core.h (hunk 5 above)
Comment (JeffreyRichter) [reply sdk/iot/core/inc/az_iot_core.h:120]: It should not be public and I think it was moved into core (maybe its still an outstanding PR?).
Maybe @ahsonkhan or @vhvb1989 know?

File: sdk/iot/doc/mqtt_state_machine.md [hunk 7]
```diff
@@ -61,14 +61,23 @@ if(az_failed(az_iot_hub_client_sas_get_signature(client, unix_time + 3600, signa
     // error.
 }
 
-// Base64Encode the HMAC256 of the az_span_ptr(signature) with the Shared Access Key.
+// Base64Encode the HMAC256 of the az_span_ptr(signature)/az_span_len(signature) with the Shared Access Key.
```
[the remark this answers is not in the document]
Comment (CIPop) [reply sdk/iot/doc/mqtt_state_machine.md:64]: Oh, I meant using those two, not divide. I see how that can be misinterpreted.
Fixed both.

File: sdk/iot/core/inc/az_iot_core.h (hunk 6 above)
Comment (CIPop) [reply sdk/iot/core/inc/az_iot_core.h:28]: It's not portable to a lot of the MQTT libraries. Most of them control clean session with a boolean flag and not the actual bit within the `CONNACK` message.
Also, it's not clear what the default meant. The guidance is to always have CleanSession=false. (see https://github.com/Azure/azure-sdk-for-c/pull/658/files#diff-19fef307f19572073d4a230f548b6b41R72 within the MQTT state machine MD file updated in this PR.)

File: sdk/iot/core/inc/az_iot_core.h (hunk 5 above)
Comment (ahsonkhan) [reply sdk/iot/core/inc/az_iot_core.h:120]: It hasn't moved to core yet. Final decision on semantics of the API was pending, which is why we decided to leave it in iot/common for now (btw, we probably want to rename iot/core to iot/common to avoid confusion).

We can (and should) move it to core in the near future (post preview?).

File: sdk/iot/core/inc/az_iot_core.h [hunk 8]
```diff
@@ -21,12 +21,16 @@
 /**
  * @brief Azure IoT service MQTT connection properties.
  *
+ * @note You MUST NOT use any symbols (macros, functions, structures, enums, etc.)
```
Comment (ahsonkhan) [inline sdk/iot/core/inc/az_iot_core.h:24]: Consider moving this `@note` to the `@file` comment block above, like other places. I don't think it belongs to this particular enum.

File: sdk/iot/core/inc/az_iot_core.h [hunk 9]
```diff
@@ -21,12 +21,16 @@
...
+ * prefixed with an underscore ('_') directly in your application code. These symbols
+ * are part of Azure SDK's internal implementation; we do not document these symbols
+ * and they are subject to change in future versions of the SDK which would break your code.
  */
+
```
Comment (ahsonkhan) [inline sdk/iot/core/inc/az_iot_core.h:29]: Remove new line?

File: sdk/iot/core/inc/az_iot_core.h [hunk 10]
```diff
@@ -58,56 +62,49 @@ typedef enum
...
-
 /**
  * @brief Checks if the status indicates a successful operation.
  *
  * @param[in] status The #az_iot_status to verify.
- * @return True if the status indicates success. False otherwise.
+ * @return `true` if the status indicates success. `false` otherwise.
  */
-AZ_NODISCARD bool az_iot_is_success_status(az_iot_status status);
+AZ_INLINE bool az_iot_is_success_status(az_iot_status status)
```
Comment (ahsonkhan) [inline sdk/iot/core/inc/az_iot_core.h:71]: Keep `AZ_NODISCARD`?

File: sdk/iot/core/inc/az_iot_core.h [hunk 11]
```diff
@@ -58,56 +62,49 @@ typedef enum
...
     int32_t operation_msec,
     int16_t attempt,
-    int32_t retry_delay_msec,
+    int32_t min_retry_delay_msec,
     int32_t max_retry_delay_msec,
     int32_t random_msec);
 
 /**
- * @brief az_span_token is a string tokenizer for az_span.
+ * @brief _az_span_token is a string tokenizer for az_span.
```
Comment (ahsonkhan) [inline sdk/iot/core/inc/az_iot_core.h:107]: Remove method name from doc to avoid redundant info and having to make changes in multiple places in the future.

```suggestion
 * @brief String tokenizer for #az_span.
```

File: sdk/iot/core/src/az_iot_core.c [hunk 12]
```diff
@@ -8,74 +8,41 @@
...
+
 #include <_az_cfg.h>
 
-AZ_NODISCARD az_result az_iot_get_status_from_uint32(uint32_t status_int, az_iot_status* status)
+AZ_NODISCARD int32_t az_iot_retry_calc_delay(
+    int32_t operation_msec,
+    int16_t attempt,
+    int32_t min_retry_delay_msec,
+    int32_t max_retry_delay_msec,
+    int32_t random_msec)
```
Comment (ahsonkhan) [inline sdk/iot/core/src/az_iot_core.c:20]: This is a relatively complex API signature, especially for something that is public. Any way to design it more simply?

Does attempt need to be an `int16_t`, or can we just keep it `int32_t` like the rest?

File: sdk/iot/core/src/az_iot_core.c [hunk 13]
```diff
@@ -8,74 +8,41 @@
...
 {
-  switch (status_int)
+  _az_PRECONDITION_RANGE(0, operation_msec, INT32_MAX - 1);
```
Comment (ahsonkhan) [inline sdk/iot/core/src/az_iot_core.c:22]: Why is `INT32_MAX` not allowed?

File: sdk/iot/core/tests/cmocka/test_az_iot_core.c [hunk 14]
```diff
@@ -157,13 +92,65 @@ static void _az_iot_u32toa_size_success(void** state)
...
+
+  assert_int_equal(
+      INT32_MAX - 1,
+      az_iot_retry_calc_delay(0, INT16_MAX - 1, INT32_MAX - 1, INT32_MAX - 1, INT32_MAX - 1));
+}
+
+#ifdef _MSC_VER
+// warning C4113: 'void (__cdecl *)()' differs in parameter lists from 'CMUnitTestFunction'
+#pragma warning(disable : 4113)
+#endif
```
Comment (ahsonkhan) [inline sdk/iot/core/tests/cmocka/test_az_iot_core.c:143]: Can you please explain exactly what about `CMUnitTestFunction` is requiring this warning disable?

File: sdk/iot/doc/mqtt_state_machine.md [hunk 15]
```diff
@@ -146,6 +155,8 @@ _Example:_
     }
 ```
 
+__Important:__ C2D messages are not enqueued until the device establishes the first MQTT session (connects for the first time to IoT Hub). The C2D message queue is preserved (according to the per-message time-to-live) as long as the device connects with Clean Session `false`.
```
Comment (ahsonkhan) [inline sdk/iot/doc/mqtt_state_machine.md:158]: What is `Clean Session`? Is it a term that folks familiar with the IoT Hub domain would be aware off, or is it part of our API surface? Consider adding a link to a reference.

File: sdk/iot/hub/inc/az_iot_hub_client.h [hunk 16]
```diff
@@ -361,7 +366,7 @@ typedef struct az_iot_hub_client_c2d_request
  * @param[out] out_request If the message is a C2D request, this will contain the
  *                         #az_iot_hub_client_c2d_request
  * @return #az_result
- *         - #AZ_ERROR_IOT_TOPIC_NO_MATCH if the topic is not matching the expected format.
+ *         - `AZ_ERROR_IOT_TOPIC_NO_MATCH` if the topic is not matching the expected format.
```
Comment (ahsonkhan) [inline sdk/iot/hub/inc/az_iot_hub_client.h:369]: Why was this change necessary? It should link to the correct `az_result` value with `#`

File: sdk/iot/hub/inc/az_iot_hub_client.h [hunk 17]
```diff
@@ -411,7 +416,7 @@ typedef struct az_iot_hub_client_method_request
  * @param[out] out_request If the message is a method request, this will contain the
  *                         #az_iot_hub_client_method_request.
  * @return #az_result
- *         - #AZ_ERROR_IOT_TOPIC_NO_MATCH if the topic is not matching the expected format.
+ *         - `AZ_ERROR_IOT_TOPIC_NO_MATCH` if the topic is not matching the expected format.
```
Comment (ahsonkhan) [inline sdk/iot/hub/inc/az_iot_hub_client.h:419]: Same here/elsewhere related to `az_result`. Keep `#`.

File: sdk/iot/provisioning/inc/az_iot_provisioning_client.h [hunk 18]
```diff
@@ -251,18 +259,63 @@ typedef struct az_iot_provisioning_client_register_response
...
+/**
+ * @brief Returns the #az_iot_provisioning_client_operation_status of a
+ * #az_iot_provisioning_client_register_response object.
+ *
+ * @param[in] response The #az_iot_provisioning_client_register_response obtained after a successful
+ *                     call to #az_iot_provisioning_client_parse_received_topic_and_payload.
+ * @param[out] out_operation_status The registration operation status.
+ * @return #az_result
+ *         - #AZ_ERROR_PARSER_UNEXPECTED_CHAR if the string contains an unexpected value.
+ *         @remark See #az_iot_provisioning_client_register_response for acceptable values.
```
Comment (ahsonkhan) [inline sdk/iot/provisioning/inc/az_iot_provisioning_client.h:295]: I don't believe this will render as intended. Move the `@remark` after the @brief in its own standalone line, if it isn't important that it be connected to and display next to the `AZ_ERROR_PARSER_UNEXPECTED_CHAR` result.

**Comment by JeffreyRichter** [comment]
Post preview is fine.


  *   Jeffrey Richter (watch Architecting Distributed Cloud Apps<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fazure.microsoft.com%2Ftraining%2Fdistributed-cloud-apps%2F&data=02%7C01%7Cjeffreyr%40microsoft.com%7C82df4ff9e5ae46a5095908d7345e957b%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637035453166233741&sdata=4GnkpFUKxlZXGXFr6oQo5GB%2FFNIbDXIZDXUij0q4AEE%3D&reserved=0>)

________________________________

File: sdk/iot/core/inc/az_iot_core.h (hunk 9 above)
Comment (CIPop) [reply sdk/iot/core/inc/az_iot_core.h:29]: Moved the entire comment in the file section.

File: sdk/iot/core/inc/az_iot_core.h (hunk 11 above)
Comment (CIPop) [reply sdk/iot/core/inc/az_iot_core.h:107]: Changed this and the one below.

File: sdk/iot/core/src/az_iot_core.c (hunk 12 above)
Comment (CIPop) [reply sdk/iot/core/src/az_iot_core.c:20]: > This is a relatively complex API signature, especially for something that is public. Any way to design it more simply?

I had a different approach that would require state. We all agreed the overheads are too big.

> Does attempt need to be an int16_t, or can we just keep it int32_t like the rest?

It's the same as the internal that Az SDKs proposed and so I can't use a larger int unless the internal also changes.

File: sdk/iot/core/src/az_iot_core.c (hunk 13 above)
Comment (CIPop) [reply sdk/iot/core/src/az_iot_core.c:22]: It's allowed but it'll generate a compile-time overflow warning and break the build.

File: sdk/iot/core/tests/cmocka/test_az_iot_core.c (hunk 14 above)
Comment (CIPop) [reply sdk/iot/core/tests/cmocka/test_az_iot_core.c:143]: It's the cmocka `void **state` parameter. I think it's not adding any value (we don't use the state feature) and it adds 3 lines of extra code.
The warning is for not using the same signature (I'm using va-args `test()` and it's expecting `test(void **)`). Using va_args (any empty () definition) is a valid C99 language feature.

File: sdk/iot/doc/mqtt_state_machine.md (hunk 15 above)
Comment (CIPop) [reply sdk/iot/doc/mqtt_state_machine.md:158]: I won't add it here though: it has its own paragraph under Connecting.

File: sdk/iot/hub/inc/az_iot_hub_client.h (hunk 16 above)
Comment (CIPop) [reply sdk/iot/hub/inc/az_iot_hub_client.h:369]: Unfortunately it doesn't and leaves `#AZ_ERROR_IOT_TOPIC_NO_MATCH` in the doc.

File: sdk/iot/hub/inc/az_iot_hub_client.h (hunk 17 above)
Comment (CIPop) [reply sdk/iot/hub/inc/az_iot_hub_client.h:419]: See above.

File: sdk/iot/provisioning/inc/az_iot_provisioning_client.h (hunk 18 above)
Comment (CIPop) [reply sdk/iot/provisioning/inc/az_iot_provisioning_client.h:295]: You're right: it renders at top-level, not within the return paragraph.

Comment (CIPop) [reply sdk/iot/provisioning/inc/az_iot_provisioning_client.h:295]: On second thought, I'm removing this completely since the value comes from a struct that we've parsed and it's unclear.

File: sdk/iot/hub/inc/az_iot_hub_client.h (hunk 16 above)
Comment (ahsonkhan) [reply sdk/iot/hub/inc/az_iot_hub_client.h:369]: I think that is because that error code isn't documented. I use `#AZ_OK` and others in `az_span.h` etc. and it works fine, because it has a `///<` comment.
